stream-chat-android-client/src/main/java/io/getstream/chat/android/client/token/TokenManagerImpl.kt

```diff
@@ -2,80 +2,89 @@ package io.getstream.chat.android.client.token
-    private val tokenProviderExecutor = Executors.newSingleThreadExecutor()
-
-    override fun loadAsync(listener: (Result<String>) -> Unit) {
-
-        if (hasToken()) {
-            listener(Result(token))
-        } else {
-            listeners.add(listener)
-            loadAsync()
-        }
-    }
+    private var job: Job? = null
 
-    override fun loadAsync() {
-        if (loadingFuture == null) {
-            loadingFuture = tokenProviderExecutor.submit {
-                loadSync()
-                loadingFuture = null
+    override fun loadAsyncAndRetry(callback: (Result<String>) -> Unit) {
+        job?.cancel()
+        job = GlobalScope.launch(DispatcherProvider.IO) {
+            val result = if (hasToken()) {
+                Result(token)
+            } else {
+                var newToken: String = runAndRetry()
+                if (newToken != EMPTY_TOKEN) {
+                    Result(newToken)
+                } else {
+                    Result(ChatError("Unable to load token"))
+                }
+            }
+            withContext(DispatcherProvider.Main) {
+                callback(result)
             }
         }
     }
 
-    override fun loadSync() {
-        try {
-            onTokenLoaded(provider.loadToken())
+    @Synchronized
+    override fun loadSync(): String {
+        return try {
+            provider.loadToken()
         } catch (t: Throwable) {
-            onTokenLoadingError(t)
+            EMPTY_TOKEN
+        }.also {
+            this.token = it
         }
     }
 
-    override fun getToken(): String {
-        return token
-    }
+    override fun getToken(): String = token
 
     override fun expireToken() {
-        token = ""
+        token = EMPTY_TOKEN
     }
 
     override fun setTokenProvider(provider: TokenProvider) {
-        this.token = ""
+        this.token = EMPTY_TOKEN
         this.provider = provider
     }
 
-    override fun hasToken(): Boolean {
-        return token.isNotEmpty()
-    }
+    override fun hasToken(): Boolean = token != EMPTY_TOKEN
 
     override fun hasTokenProvider(): Boolean {
         return this::provider.isInitialized
     }
 
-    /**
-     * This method is synchronized because this class is not thread safe. Without
-     * synchronization concurrent invocations of [loadAsync] and [loadSync] can lead
-     * to the state when [listeners] are notified several times.
-     */
-    @Synchronized
-    private fun onTokenLoaded(token: String?) {
-        this.token = token ?: ""
-        listeners.forEach { it?.invoke(Result(this.token)) }
-        listeners.clear()
+    override fun shutdown() {
+        job?.cancel()
+    }
+
+    private suspend fun runAndRetry(): String {
+        var attempt = 0
+        var token: String = EMPTY_TOKEN
+
+        while (attempt < ATTEMPT_COUNT) {
+            token = loadSync()
+            if (token != EMPTY_TOKEN) {
+                break
+            } else {
+                delay(RETRY_DELAY_MILLIS)
```

Comment:
Could we use an Exponential backoff?

Reply:
Yes, will do. And later I will review our retry use cases, maybe we can come up with a neat solution.